src/Database/LSMTree/Internal.hs

```diff
@@ -129,6 +131,18 @@ data Session' m = forall h. Typeable h => Session' !(Session m h)
 instance NFData (Session' m) where
   rnf (Session' s) = rnf s
 
+type Table' :: (Type -> Type) -> Type -> Type -> Type -> Type
+data Table' m k v b = forall h. Typeable h => Table' (Table m h)
+
+instance NFData (Table' m k v b) where
+  rnf (Table' t) = rnf t
```

Comment:
I'm not a big fan of the pattern of defining these types here and using synonyms in the API modules, as synonyms can leak into error messages and it also makes the diff between them larger vs having three `data Table` there.

But the pattern already exists, so it makes sense to follow it here. :+1:

Reply:
In our current setup, we need to expose the constructors for our `NoThunks` assertions, but @dcoutts suggested we do not expose the constructors from the public API. That's why I moved them here and added type synonyms in the public APIs

Reply:
IMO it would have been fine to expose the constructors in the public API, since if you want to do anything with the internal `Table` then you'll have to import `Internal` modules anyway